Instrumenting cassandra executeReactive method

SimoneGiusso opened the pull request:
It follows the [issue](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/6395#issue-1323561263) I opened some days ago.

The `executeReactive` method use the same processor used by `executeAsync` (see [here](https://github.com/datastax/java-driver/blob/65d2c19c401175dcc6c370560dd5f783d05b05b9/core/src/main/java/com/datastax/dse/driver/internal/core/cql/reactive/CqlRequestReactiveProcessor.java#L53)) and wrap the callback in the `DefaultReactiveResultSet` publisher.

Here I'm simply overriding the `executeReactive` method doing the same thing: call the already instrumented `executeAsync` method and wrapping the callback using the `DefaultReactiveResultSet` publisher.

~~I did an upgrade of the `java-driver-core` library to have `TracingCqlSession.java` extending the `ReactiveSession`. I have to probably rename the `cassandra-4.0` module in `cassandra-4.14` but I'll let you confirm this.~~ -> Cassandra-4.4 is enough.

trask commented:
thx @SimoneGiusso!

> I have to probably rename the `cassandra-4.0` module in `cassandra-4.14` but I'll let you confirm this.

we probably don't want to drop support for earlier cassandra 4.x versions, see our reasoning under the "Javaagent instrumentation" section here:
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/VERSIONING.md#dropping-support-for-older-library-versions

instead, go ahead and create a new module `cassandra-4.14` that contains just the reactive instrumentation. it's a little odd to have overlapping instrumentation modules (4.0 will still apply to 4.14), but we have a couple of other cases like this (e.g. mongo) and it's the best approach we have come up with to handle this kind of case

SimoneGiusso commented:
I see. I'll probably able to do this change in two weeks, when I'll be back from vacations (I won't bring my laptop with me).

SimoneGiusso commented:
Some tests are not passing but I'm missing why. Locally everything seems ok. Same for muzzle. The output of these commands:

- ./gradlew :instrumentation:cassandra:cassandra-4.14:javaagent:test --tests CassandraClientTest
- ./gradlew muzzle

Don't show any error.

SimoneGiusso commented on instrumentation/cassandra/cassandra-4.0/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_0/CassandraSingletons.java:
```diff
@@ -32,8 +31,6 @@ public final class CassandraSingletons {
             .addAttributesExtractor(
                 SqlClientAttributesExtractor.builder(attributesGetter)
                     .setTableAttribute(SemanticAttributes.DB_CASSANDRA_TABLE)
-                    .setStatementSanitizationEnabled(
```
This is just to have the build phase successfully passing, to avoid:

> package io.opentelemetry.javaagent.bootstrap.internal does not exist import io.opentelemetry.javaagent.bootstrap.internal.CommonConfig;

Currently I don't have a better solution than create a full copy of the `TracingCqlSession.java` (this class is used here) in the `cassandra.v4_14` module and move the "old" `TracingCqlSession.java` from `library` module to `javaagent` module (together with `CassandraSingletons`)

mateuszrzeszutek replied:
That error happens because the library modules cannot access javaagent code.
Instead of just copying the `CassandraSingletons` class, could you add a `CassandraTelemetry` & `CassandraTelemetryBuilder` classes that allow programatic configuration? You can take a look at the mongodb [telemetry](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/MongoTelemetry.java) and [builder](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/MongoTelemetryBuilder.java) classes for an example.

SimoneGiusso replied:
I'm not seeing the solution here. It's not just a problem related to `CassandraSingletons`... `CassandraSingletons` contains [jaavagent](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/4cbfd7da8b80d4f76e56537279f1f15a012ad51a/instrumentation/cassandra/cassandra-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_0/CassandraSingletons.java#L15) code. If we remove this dependency by adding (for example) a boolean parameter for the [instrumenter()](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/4cbfd7da8b80d4f76e56537279f1f15a012ad51a/instrumentation/cassandra/cassandra-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_0/CassandraSingletons.java#L44) method there is still the problem that we need to pass this parameter (this `CommonConfig.get().isStatementSanitizationEnabled()`) in the `TracingCqlSession` class every time we call `instrumenter()`.

However this `TracingCqlSession` is part of the `library` module (i.e. we cannot have this code `instrumenter(CommonConfig.get().isStatementSanitizationEnabled())`) otherwise we still have the same problem.

We could add a boolean paramenter in the `TracingCqlSession` and finally configure this boolean flag in the [CompletionStageFunction](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/4cbfd7da8b80d4f76e56537279f1f15a012ad51a/instrumentation/cassandra/cassandra-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_0/CompletionStageFunction.java#L22) (it's javaagent code): in other words pass the configuration in the TracingCqlSession constructor. Otherwise we still need to move `TracingCqlSession` in the javaagent module.

Do you agree or I didn't get your suggestion?

mateuszrzeszutek replied:
What I meant was that `CassandraSingletons` should not be present in the library instrumentation module at all. The `*Singletons` classes are (by convention) only used in javaagent instrumentations, where you need to have some sort of a global var/instance.

Consider the following example/pseudocode:
```java
class CassandraTelemetry {
  static CassandraTelemetryBuilder(OpenTelemetry otel) {
    return new CassandraTelemetryBuilder(otel);
  }
  
  private final Instrumenter<CassandraRequest, ExecutionInfo> instrumenter;
  private final boolean statementSanitizationEnabled;
  // constructor

  public CqlSession decorate(CqlSession original) {
    return new TracingCqlSession(instrumenter, statementSanitizationEnabled, original);
  }
}

class CassandraTelemetryBuilder {
  CassandraTelemetryBuilder setStatementSanitizationEnabled(boolean enabled) {
    // ...
  }
  
  CassandraTelemetry build() {
    // construct the Instrumenter
    return new CassandraTelemetry(instrumenter, statementSanitizationEnabled);
  }
}
```

You're not using `CommonConfig` anywhere, you're allowing the user of `CassandraTelemetryBuilder` to pass the option programmatically through a builder method. The javaagent `CassandraSingletons` class could then look just like that:

```java
class CassandraSingletons {
  static final CassandraTelemetry telemetry = CassandraTelemetry.builder(GlobalOpenTelemetry.get())
      .setStatementSanitizationEnabled(CommonConfig....)
      .build();
}
```

SimoneGiusso replied:
Thanks, it is clearer now. I [committed](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/128f0f620b863a774d02ded395bd8067ff2e6030) this changes.

SimoneGiusso commented on instrumentation/cassandra/cassandra-4.14/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_14/TracingCqlSession.java:
```diff
@@ -0,0 +1,24 @@
+    super(session);
+  }
+
+  @Override
+  public ReactiveResultSet executeReactive(Statement<?> statement) {
+    return new DefaultReactiveResultSet(() -> executeAsync(statement));
+  }
```
This is the core change.

SimoneGiusso commented on instrumentation/cassandra/cassandra-4.0/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_0/CassandraTelemetry.java:
```diff
@@ -0,0 +1,53 @@
+        .buildInstrumenter(SpanKindExtractor.alwaysClient());
+  }
+
+  public Instrumenter<CassandraRequest, ExecutionInfo> getInstrumenter() {
+    return instrumenter;
+  }
```
I think I didn't do that because how can we use the `TracingCqlSession` with the reactive method'? In this way the `CassandraTelemetry` (shared in both javaagent module: cassandra4.0 and cassandra4.4) will decorate only the original TracingCqlSession.

SimoneGiusso replied:
Maybe this is not needed if I simply wrap the original tracingCql. I did a [clean-up](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/39b3d728cd6051d56c6796b05e532f988f675190).

SimoneGiusso replied:
[Done](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/1895bfca1ae5ebc26ca091aab7bbad9032b6fdb4)!

SimoneGiusso commented on instrumentation/cassandra/cassandra-4.0/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_0/CassandraTelemetry.java:
```diff
@@ -0,0 +1,53 @@
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+
+public final class CassandraTelemetry {
+
```
I simply skip it since is not used. I'm not a fan of having unused code but I'll add it.

SimoneGiusso replied:
Maybe this is not needed if I simply wrap the original tracingCql. I did a [clean-up](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/39b3d728cd6051d56c6796b05e532f988f675190).

SimoneGiusso replied:
[Done](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/1895bfca1ae5ebc26ca091aab7bbad9032b6fdb4)!.

SimoneGiusso commented on instrumentation/cassandra/cassandra-4.0/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_0/CassandraTelemetryBuilder.java:
```diff
@@ -0,0 +1,36 @@
+    this.openTelemetry = openTelemetry;
+  }
+
+  public CassandraTelemetryBuilder setInstrumentationName(String instrumentationName) {
```
Sorry I'm not understanding here. I saw that this `KafkaInstrumenterFactory` is like a builder and it is receiving the instrumentationName with the constructor. Where should it be used? 

- Not in the `CassandraTelemetryBuilder` otherwise we should pass the instrumentation_name name as parameter (then what is the sense of having the Factory class?).
- Not in the singleton (where `KafkaInstrumenterFactory` is used) instead of the builder, otherwise what is the sense of having the builder?

SimoneGiusso replied:
I left unchanged the `cassandra-4.0` and in `cassandra-4.4` the instrumentation name cannot be set (it's defaulted).

trask commented:
hi @SimoneGiusso! just checking in to see if you are still working on this? (looks like a great addition)

definitely let us know if you have any questions, thx!

hansh0801 commented:
track

SimoneGiusso commented:
> hi @SimoneGiusso! just checking in to see if you are still working on this? (looks like a great addition)
> 
> definitely let us know if you have any questions, thx!

Sorry, coming back to this by starting addressing some comments from @mateuszrzeszutek. This PR is getting a bit bigger when the actual "feature" is just one method with one line of code 😀. However I'll try my best to finalize this soon.

SimoneGiusso commented on instrumentation/cassandra/cassandra-4.14/javaagent/build.gradle.kts:
```diff
@@ -0,0 +1,29 @@
+  pass {
+    group.set("com.datastax.oss")
+    module.set("java-driver-core")
+    versions.set("[4.14,)")
```
Done -> https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/3bc43205aaa5de51ed0c951202d0dee4bfb756f2

SimoneGiusso commented on instrumentation/cassandra/cassandra-4.14/javaagent/src/test/groovy/CassandraClientTest.groovy:
```diff
@@ -0,0 +1,129 @@
+
+import static io.opentelemetry.api.trace.SpanKind.CLIENT
+
+class CassandraClientTest extends AgentInstrumentationSpecification {
```
I'm creating a common abstract base class as suggested. When running `./gradlew :instrumentation:cassandra:cassandra-4.0:javaagent:test --tests CassandraClientTest` I'm getting `java.lang.ClassNotFoundException: io.opentelemetry.instrumentation.cassandra.AbstractCassandraClientTest`. @trask @mateuszrzeszutek can you help me understand why the class loader is not finding this class?

SimoneGiusso replied:
[Done](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/15172a46789fe31e155932bd94524affd7faac64)!

trask commented on instrumentation/cassandra/testing/build.gradle.kts:
```diff
@@ -0,0 +1,16 @@
+plugins {
+  id("otel.javaagent-instrumentation")
+}
+
+dependencies {
+  api(project(":testing-common"))
+
+  library("org.testcontainers:testcontainers:1.17.5")
+  library("com.datastax.oss:java-driver-core:4.4.0")
+}
```
```suggestion
plugins {
  id("otel.java-conventions")
}

dependencies {
  api(project(":testing-common"))

  implementation("org.testcontainers:testcontainers:1.17.5")
  implementation("com.datastax.oss:java-driver-core:4.4.0")
}
```

trask reviewed (commented):
hey @SimoneGiusso,

the change below should get the tests running

SimoneGiusso commented on instrumentation/cassandra/cassandra-4.14/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_14/TracingCqlSession.java:
```diff
@@ -0,0 +1,28 @@
+import io.opentelemetry.instrumentation.cassandra.v4_0.CassandraRequest;
+
+public class TracingCqlSession
+    extends io.opentelemetry.instrumentation.cassandra.v4_0.TracingCqlSession {
```
[Done](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/15172a46789fe31e155932bd94524affd7faac64)!

SimoneGiusso commented on instrumentation/cassandra/cassandra-4.0/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_0/CassandraTelemetry.java:
```diff
@@ -0,0 +1,53 @@
+        createInstrumenter(openTelemetry, instrumentationName, statementSanitizationEnabled);
+  }
+
+  private static Instrumenter<CassandraRequest, ExecutionInfo> createInstrumenter(
```
Maybe this is not needed if I simply wrap the original tracingCql. I did a [clean-up](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/39b3d728cd6051d56c6796b05e532f988f675190).

SimoneGiusso replied:
[Done](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/1895bfca1ae5ebc26ca091aab7bbad9032b6fdb4)!

SimoneGiusso commented on instrumentation/cassandra/cassandra-4.0/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_0/CassandraTelemetryBuilder.java:
```diff
@@ -0,0 +1,36 @@
+
+import io.opentelemetry.api.OpenTelemetry;
+
+public class CassandraTelemetryBuilder {
```
Maybe this is not needed if I simply wrap the original tracingCqlsession. I did a [clean-up](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/39b3d728cd6051d56c6796b05e532f988f675190).

SimoneGiusso commented on instrumentation/cassandra/cassandra-4.0/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_0/CassandraTelemetryBuilder.java:
```diff
@@ -0,0 +1,36 @@
+  private String instrumentationName = DEFAULT_INSTRUMENTATION_NAME;
+  private boolean statementSanitizationEnabled;
+
+  public CassandraTelemetryBuilder(OpenTelemetry openTelemetry) {
```
Maybe this is not needed if I simply wrap the original tracingCqlsession. I did a [clean-up](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/39b3d728cd6051d56c6796b05e532f988f675190).

SimoneGiusso replied:
[Done](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/1895bfca1ae5ebc26ca091aab7bbad9032b6fdb4)!

SimoneGiusso commented on instrumentation/cassandra/cassandra-4.0/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_0/CassandraTelemetry.java:
```diff
@@ -0,0 +1,53 @@
+import io.opentelemetry.instrumentation.api.instrumenter.net.NetClientAttributesExtractor;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+
+public final class CassandraTelemetry {
```
Maybe this is not needed if I simply wrap the original tracingCqlsession. I did a [clean-up](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/39b3d728cd6051d56c6796b05e532f988f675190).

SimoneGiusso replied:
[Done](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/1895bfca1ae5ebc26ca091aab7bbad9032b6fdb4)! I added some javadoc in the `*Builder` class too.

SimoneGiusso commented on instrumentation/cassandra/cassandra-4.14/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_14/CassandraClientInstrumentationModule.java:
```diff
@@ -0,0 +1,25 @@
+  public CassandraClientInstrumentationModule() {
+    super("cassandra", "cassandra-4.14");
+  }
+
```
[Done](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/9879dd7db4cac67a47794d28e10e878a8f9f6a97)!

SimoneGiusso commented:
Hi @trask and @mateuszrzeszutek. To sum-up: we are instrumenting the `executeReactive` method (introduced in Cassandra 4.4). It consists to "simply" enrich the current `TracingCqlSession` already implemented and have a new test for this method.

We have decided to not modify the current `cassandra-4.0` module but to create a new `cassandra-4.4` module. For this reason a new testing module has been created which groups all the common tests between these two modules (i.e. all the tests in the current `cassandra-4.0`). The new module entry-point is `CassandraTelemetry` (following the @mateuszrzeszutek's suggestions). A bit of classes has been duplicated since they are package protected (e.g `CassandraAttributesExtractor`). Could we maybe need to create a common package?

To reuse the original `TracingCqlSession` a small modification has been done, a new constructor was created which accepts the instrumenter (the instrumenter is different between 4.0 and 4.4, just for the name). The original constructor should be kept since this class (and the constructor) is public and could be used as entry point in some users' projects. 

Can `cassandra-4.0` considered as deprecated since `cassandra-4.4` have all the features of `cassandra-4.0` + reactive session? Users should probably start use `cassandra-4.4` instead.

There is one last point I left open in the code. Since `cassandra-4.4` depends on `cassandra-4.0` in `*.v4_4.CompletionStageFunction.java` the original `TracingCqlSession` is injected instead of `DefaultSession` (I left a comment there with a "workaround" method). This is needed to be handle differently.

SimoneGiusso commented on instrumentation/cassandra/cassandra-4.4/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_4/TracingCqlSession.java:
```diff
@@ -0,0 +1,176 @@
+  }
+
+  @Override
+  public ReactiveResultSet executeReactive(Statement<?> statement) {
+    return new DefaultReactiveResultSet(() -> originalTracingCqlSession.executeAsync(statement));
+  }
```
This is the core change to instrument the method.

SimoneGiusso commented:
Hi @trask and @mateuszrzeszutek. 

To sum-up: we are instrumenting the `executeReactive` method (introduced in Cassandra 4.4). It consists in "simply" enrich the current `*.v4_0.TracingCqlSession` already implemented and add a new test for this method.

We have decided to not modify the current `cassandra-4.0` module but to create a new `cassandra-4.4` module. For this reason a new testing module has been created which groups all the common tests between these two modules (i.e. all the tests in the current `cassandra-4.0`). 

The new entry point for both modules is the `CassandraTelemetry` class (following the @mateuszrzeszutek's suggestions). Two distinct classes are needed, one for each module, to create two distinct `TracingCqlSession` objects (see `CassandraTelemetry.wrap()`. Inheritance here has been used to:

- Set the instrumenter's name.
- Share the code to create the instrumenter (and avoid duplicating package private classes such as `CassandraAttributesExtractor`).

The new `TracingCqlSession` simply wrap the original `TracingCqlSession` and enables tracing of the `executeReactive` method. In other words `cassandra-4.4` is offering what is already offered by `cassandra-4.0`, with the addition of this `executeReactive` method.

Could you please review the PR and possibly ask questions if needed? Thanks.

hansh0801 commented:
track

SimoneGiusso commented:
Hi, @trask @mateuszrzeszutek. Did you have some time to check this PR. I think we are close to merge this branch. All tests are passing and I should have applied almost all your suggestions.

trask commented:
hi @SimoneGiusso sorry for the long delay, we will try to get to this in the next week

trask commented:
hey @SimoneGiusso! I pushed some changes to your branch:

* merged in the new cassandra Java tests
* added testing for the library instrumentation
* changed library instrumentation to target 4.4 instead of 4.0 (we generally want library instrumentation to target the latest, though this does mean now there's some code duplication between the 4.0 and 4.4 implementation, I think this is ok, I don't think we currently have a pattern for sharing code between a later library version and an earlier javaagent version)

trask commented on instrumentation/cassandra/cassandra-4.4/testing/src/main/java/io/opentelemetry/testing/cassandra/v4_4/AbstractCassandra44Test.java:
```diff
@@ -0,0 +1,8 @@
+
+public abstract class AbstractCassandra44Test extends AbstractCassandraTest {
+
+  // TODO add reactive tests here
```
can you rewrite the couple of reactive groovy tests here in Java?

SimoneGiusso replied:
https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6441/commits/2385797ceceaa103e0daaadceaafda5f6246ea75

trask commented on instrumentation/cassandra/cassandra-4.4/library/src/main/java/io/opentelemetry/instrumentation/cassandra/v4_4/CassandraAttributesExtractor.java:
```diff
@@ -0,0 +1,79 @@
+        SemanticAttributes.DB_CASSANDRA_SPECULATIVE_EXECUTION_COUNT,
+        executionInfo.getSpeculativeExecutionCount());
+
+    // TODO can this be rewritten using newer API?
+    @SuppressWarnings("deprecation")
+    Statement<?> statement = executionInfo.getStatement();
```
@SimoneGiusso can you look at this to see if there's a better way to do this in 4.4+

SimoneGiusso replied:
They [made `ExecutionInfo` compatible with any `Request` type](https://github.com/datastax/java-driver/pull/1367/files), and the current deprecated method just do a cast. We can do the [same](https://github.com/datastax/java-driver/blob/355844408325df5045fc5b10d99737f24003042d/core/src/main/java/com/datastax/oss/driver/internal/core/cql/Conversions.java#L324).

trask commented:
it would be great to add a README for the new library instrumentation, see #6947, but we can merge without it too

trask commented:
thx @SimoneGiusso!

SimoneGiusso commented:
Happy to contribute. Thank you for your help and reviews @trask!
